File: Python/20.py
class Solution:
    # @param {string} s
    # @return {boolean}
    def isValid(self, s):
        stack = []
        dic = {')':'(',']':'[','}':'{'}
        for i in s:
            if i in ('(', '[', '{'):
                stack.append(i)
            else:
                if stack and dic[i] == stack[-1]:
                    del stack[-1]
                else:
                    return False
        return True if not stack else False
                

File: C++/20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack <char> test;
        map <char,char> table = {{'(',')'},{'[',']'},{'{','}'}};
        for(int i=0; i<s.size();i++){
            if(s[i] == ')' || s[i]==']'||s[i]=='}'){
                if(test.empty() || table[test.top()]!=s[i]) return false;
                test.pop();
            }
            else test.push(s[i]);
        }
        return test.empty; //Awesome! test.empty()?true:false; 
    }
};